### PR TITLE
Stop and custom words remover to estimator

### LINKED ISSUE #725: CustomStopWordsRemoverTransform shouldn't hard-code its column name
See the conversation here:
https://github.com/dotnet/machinelearning/pull/700/files/e30785ff454627c7b74e3b6873d70b9e6de28bdf#r212152727

image

We should consider making the column name configurable instead of hard-coding the column name, which may be hiding an existing column.

/cc @codemzs

### PR DESCRIPTION
**Pull request by Ivanidzo4ka**
Convert stop word remover and custom stop word remover to estimator/transformer.
Fix #725

### CODE REVIEW

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 1]
```diff
@@ -286,121 +178,51 @@ private static NormStr.Pool[] StopWords
             }
         }
 
-        public StopWordsRemovingTransformer(IHostEnvironment env, Arguments args, IDataView input)
-            : base(env, RegistrationName, Contracts.CheckRef(args, nameof(args)).Column, input, TestIsTextVector)
+        public class ColumnInfo
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: >ColumnInfo [](start = 21, length = 10)

Add comment.

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 2]
```diff
@@ -236,18 +132,14 @@ private static VersionInfo GetVersionInfo()
                 verReadableCur: 0x00010001,
                 verWeCanReadBack: 0x00010001,
                 loaderSignature: LoaderSignature,
-                loaderAssemblyName: typeof(StopWordsRemovingTransformer).Assembly.FullName);
+                loaderAssemblyName: typeof(StopWordsRemoverTransform).Assembly.FullName);
         }
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: add new line

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 3]
```diff
@@ -529,14 +360,235 @@ private static void AddResourceIfNotPresent(Language lang)
...
+        private sealed class Mapper : MapperBase
+        {
+            private readonly ColumnType[] _types;
+            private readonly StopWordsRemoverTransform _parent;
+            private readonly int[] _languageColumns;
+            private readonly bool?[] _resourcesExist;
+            private readonly Dictionary<int, int> _colMapNewToOld;
+
+            public Mapper(StopWordsRemoverTransform parent, Schema inputSchema)
+             : base(parent.Host.Register(nameof(Mapper)), inputSchema)
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: >parent [](start = 20, length = 6)

CheckRef

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 4]
```diff
@@ -529,14 +360,235 @@ private static void AddResourceIfNotPresent(Language lang)
...
+
+                        var srcValues = src.GetValues();
+                        for (int i = 0; i < srcValues.Length; i++)
+                        {
+                            if (srcValues[i].IsEmpty)
+                                continue;
+                            buffer.Clear();
+                            ReadOnlyMemoryUtils.AddLowerCaseToStringBuilder(srcValues[i].Span, buffer);
+
+                            // REVIEW nihejazi: Consider using a trie for string matching (Aho-Corasick, etc.)
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: >nihejazi [](start = 38, length = 8)

remove aliases

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 5]
```diff
@@ -529,14 +360,235 @@ private static void AddResourceIfNotPresent(Language lang)
...
+#pragma warning disable MSML_GeneralName // These names correspond to file names, so this is fine in this case.
+            Portuguese_Brazilian = 9,
+            Norwegian_Bokmal = 10,
+#pragma warning restore MSML_GeneralName
+            Russian = 11,
+            Polish = 12,
+            Czech = 13,
+            Arabic = 14,
+            Japanese = 15
+        }
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: add new line

**Comment by Ivanidzo4ka** [comment]
>         public const string LoaderSignature = "CustomStopWords";

internal

---
Refers to: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs:651 in 5e1fcae. [](commit_id = 5e1fcaefc85c9b489a055c6970d6ced80f9bf0e8, deletion_comment = False)

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 6]
```diff
@@ -777,55 +800,90 @@ private void LoadStopWords(IHostEnvironment env, IChannel ch, ArgumentsBase load
                 ch.CheckUserArg(stopWordsMap.Count > 0, nameof(Arguments.DataFile), "dataFile is empty");
             }
         }
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: add new line

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 7]
```diff
@@ -858,102 +916,170 @@ private CustomStopWordsRemovingTransformer(IHost host, ModelLoadContext ctx, IDa
...
-            ctx.CheckAtModel();
-            ctx.SetVersionInfo(GetVersionInfo());
+            private readonly ColumnType[] _types;
+            private readonly CustomStopWordsRemoverTransform _parent;
 
-            // *** Binary format ***
-            // <base>
-            SaveBase(ctx);
+            public Mapper(CustomStopWordsRemoverTransform parent, Schema inputSchema)
+             : base(parent.Host.Register(nameof(Mapper)), parent, inputSchema)
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: >parent [](start = 20, length = 6)

CheckRef

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 8]
```diff
@@ -858,102 +916,170 @@ private CustomStopWordsRemovingTransformer(IHost host, ModelLoadContext ctx, IDa
...
                 {
-                    Host.CheckValue(c, nameof(ctx));
-                    c.CheckAtModel();
-                    c.SetVersionInfo(GetStopwrodsManagerVersionInfo());
+                    InputSchema.TryGetColumnIndex(_parent.ColumnPairs[i].input, out int colIndex);
+                    Host.Assert(colIndex >= 0);
+                    result[i] = new Schema.Column(_parent.ColumnPairs[i].output, _types[i], null);
+                }
+                return result;
+            }
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: > } [](start = 11, length = 2)

add new line

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 9]
```diff
@@ -858,102 +916,170 @@ private CustomStopWordsRemovingTransformer(IHost host, ModelLoadContext ctx, IDa
...
+    public sealed class CustomStopWordsRemoverEstimator : TrivialEstimator<CustomStopWordsRemoverTransform>
+    {
 
-                        // REVIEW nihejazi: Consider using a trie for string matching (Aho-Corasick, etc.)
-                        if (_stopWordsMap.Get(buffer) == null)
-                            list.Add(srcValues[i]);
-                    }
+        public static bool IsColumnTypeValid(ColumnType type) => type.ItemType.IsText && type.IsVector;
+
+        internal const string ExpectedColumnType = "vector of Text type";
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: Do I even use them?

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 10]
```diff
@@ -20,33 +21,35 @@
 using System.Text;
 using System.Threading;
 
-[assembly: LoadableClass(StopWordsRemovingTransformer.Summary, typeof(StopWordsRemovingTransformer), typeof(StopWordsRemovingTransformer.Arguments), typeof(SignatureDataTransform),
+[assembly: LoadableClass(StopWordsRemoverTransform.Summary, typeof(IDataTransform), typeof(StopWordsRemoverTransform), typeof(StopWordsRemoverTransform.Arguments), typeof(SignatureDataTransform),
```
Comment (sfilipi) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: >StopWordsRemoverTransform [](start = 25, length = 25)

I think we want them with a verb in the name. (Removing)

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 11]
```diff
@@ -286,121 +178,51 @@ private static NormStr.Pool[] StopWords
...
-                ch.Warning(wrnMsg, missings.ToString());
-            }
+            var type = inputSchema.GetColumnType(srcCol);
+            if (!StopWordsRemoverEstimator.IsColumnTypeValid(type))
+                throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", ColumnPairs[col].input, StopWordsRemoverEstimator.ExpectedColumnType, type.ToString());
         }
 
-        public static StopWordsRemovingTransformer Create(IHostEnvironment env, ModelLoadContext ctx, IDataView input)
+        /// <summary>
+        /// Stopword remover removes language-specific lists of stop words (most common words).
```
Comment (sfilipi) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: >lists [](start = 55, length = 5)

I'd put it as "list"

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 12]
```diff
@@ -286,121 +178,51 @@ private static NormStr.Pool[] StopWords
...
-            ctx.CheckAtModel(GetVersionInfo());
-            env.CheckValue(env, nameof(env));
-            env.CheckValue(input, nameof(input));
-
-            var h = env.Register(RegistrationName);
-            return h.Apply("Loading Model", ch => new StopWordsRemovingTransformer(h, ctx, input));
+            _columns = columns;
         }
 
         public override void Save(ModelSaveContext ctx)
```
Comment (sfilipi) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: >Save [](start = 29, length = 4)

comment, if feeling generous.

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 13]
```diff
@@ -286,121 +180,54 @@ private static NormStr.Pool[] StopWords
...
+        /// Describes how the transformer handles one column pair.
+        /// </summary>
+        public class ColumnInfo
         {
-            Host.AssertNonEmpty(Infos);
-            Host.Assert(Infos.Length == Utils.Size(args.Column));
+            public readonly string Input;
+            public readonly string Output;
+            public readonly StopWordsRemovingEstimator.Language Language;
+            public readonly string LanguageColumn;
```
Comment (sfilipi) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: do those need to be public?

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 14]
```diff
@@ -286,121 +180,54 @@ private static NormStr.Pool[] StopWords
...
 
-            using (var ch = Host.Start("construction"))
+            /// <summary>
+            /// Describes how the transformer handles one column pair.
+            /// </summary>
+            /// <param name="input">Name of input column.</param>
+            /// <param name="output">Name of output column.</param>
+            /// <param name="language">Language-specific stop words list.</param>
+            /// <param name="languageColumn">Optional column to use for languages. This overrides language value.</param>
+            public ColumnInfo(string input, string output, StopWordsRemovingEstimator.Language language = StopWordsRemovingEstimator.Defaults.DefaultLanguage, string languageColumn = null)
```
Comment (sfilipi) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: nullable?

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 15]
```diff
@@ -529,14 +365,241 @@ private static void AddResourceIfNotPresent(Language lang)
...
+        /// </summary>
+        /// <param name="env">The environment.</param>
+        /// <param name="columns">Pairs of columns to remove stop words on.</param>
+        /// <param name="language">Langauge of the input text columns <paramref name="columns"/>.</param>
+        public StopWordsRemovingEstimator(IHostEnvironment env, (string input, string output)[] columns, Language language = Language.English)
+            : this(env, columns.Select(x => new StopWordsRemovingTransform.ColumnInfo(x.input, x.output, language)).ToArray())
+        {
+        }
+
+        public StopWordsRemovingEstimator(IHostEnvironment env, params StopWordsRemovingTransform.ColumnInfo[] columns)
```
Comment (sfilipi) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: >publi [](start = 8, length = 5)

add comment here too.

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 16]
```diff
@@ -602,32 +668,31 @@ private static VersionInfo GetVersionInfo()
                 verReadableCur: 0x00010001,
                 verWeCanReadBack: 0x00010001,
                 loaderSignature: LoaderSignature,
-                loaderAssemblyName: typeof(CustomStopWordsRemovingTransformer).Assembly.FullName);
+                loaderAssemblyName: typeof(CustomStopWordsRemovingTransform).Assembly.FullName);
         }
 
-        public const string StopwordsManagerLoaderSignature = "CustomStopWordsManager";
-        private static VersionInfo GetStopwrodsManagerVersionInfo()
+        private const string StopwordsManagerLoaderSignature = "CustomStopWordsManager";
```
Comment (sfilipi) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: >StopwordsManagerLoaderSignature [](start = 29, length = 31)

do you even want to define a variable for it.. the compiler will probably optimize it away.

File: src/Microsoft.ML.Transforms/Text/TextFeaturizingEstimator.cs [hunk 17]
```diff
@@ -345,23 +338,21 @@ public ITransformer Fit(IDataView input)
...
-                    col.Language = tparams.StopwordsLanguage;
+                    var tempName = GenerateColumnName(view.Schema, wordTokCols[i], "StopWordsRemoverTransform");
+                    var col = new StopWordsRemovingTransform.ColumnInfo(wordTokCols[i], tempName, tparams.StopwordsLanguage);
+                    dstCols[i] = tempName;
+                    tempCols.Add(tempName);
 
                     xfCols[i] = col;
                 }
-                view = tparams.StopWordsRemover.CreateComponent(h, view, xfCols);
+                view = new StopWordsRemovingEstimator(h, xfCols).Fit(view).Transform(view);
```
Comment (Zruty0) [inline src/Microsoft.ML.Transforms/Text/TextFeaturizingEstimator.cs]: >StopWordsRemovingEstimator [](start = 27, length = 26)

looks like this was the last one, right? Can we now convert `TextFeaturizingTransformer` to become a chain?

Comment (Zruty0) [reply src/Microsoft.ML.Transforms/Text/TextFeaturizingEstimator.cs]: not in this PR of course


---
In reply to: [235548815](https://github.com/dotnet/machinelearning/pull/1646#discussion_r235548815) [](ancestors = 235548815)

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 18]
```diff
@@ -858,102 +921,168 @@ private CustomStopWordsRemovingTransformer(IHost host, ModelLoadContext ctx, IDa
...
-                    }
+        /// <summary>
+        /// Removes stop words from incoming token streams in <paramref name="inputColumn"/>
+        /// and outputs the token streams without stopwords as <paramref name="outputColumn"/>.
+        /// </summary>
+        /// <param name="env">The environment.</param>
+        /// <param name="inputColumn">The column containing text to remove stop words on.</param>
+        /// <param name="outputColumn">The column containing output text. Null means <paramref name="inputColumn"/> is replaced.</param>
+        /// <param name="stopwords">Array of words to remove.</param>
+        public CustomStopWordsRemovingEstimator(IHostEnvironment env, string inputColumn, string outputColumn = null, params string[] stopwords)
```
Comment (Zruty0) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: >params [](start = 118, length = 6)

I don't think `params` is useful here

File: src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs [hunk 19]
```diff
@@ -858,102 +921,168 @@ private CustomStopWordsRemovingTransformer(IHost host, ModelLoadContext ctx, IDa
...
-                    VBufferUtils.Copy(list, ref dst, list.Count);
-                };
+        /// <summary>
+        /// Removes stop words from incoming token streams in input columns
+        /// and outputs the token streams without stop words as output columns.
+        /// </summary>
+        /// <param name="env">The environment.</param>
+        /// <param name="columns">Pairs of columns to remove stop words on.</param>
+        /// <param name="stopwords">Array of words to remove.</param>
+        public CustomStopWordsRemovingEstimator(IHostEnvironment env, (string input, string output)[] columns, string[] stopwords) :
```
Comment (Zruty0) [inline src/Microsoft.ML.Transforms/Text/StopWordsRemovingTransformer.cs]: >stopwords [](start = 120, length = 9)

camelCase to `stopWords`

**Pull request merged**